drivers/network/ndis/ndis/miniport.c

```diff
@@ -2335,6 +2335,71 @@ NdisIDeviceIoControl(
   return Status;
 }
 
+static
+NTSTATUS
+NdisIRemoveDevice(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _In_ PIRP Irp)
+{
+    NTSTATUS Status;
+    PLOGICAL_ADAPTER Adapter = DeviceObject->DeviceExtension;
+    
+    if (Adapter->NdisMiniportBlock.SymbolicLinkName.Buffer)
+    {
+        IoSetDeviceInterfaceState(&Adapter->NdisMiniportBlock.SymbolicLinkName, FALSE);
+        RtlFreeUnicodeString(&Adapter->NdisMiniportBlock.SymbolicLinkName);
+    }
+
+    Irp->IoStatus.Status = STATUS_SUCCESS;
+    IoSkipCurrentIrpStackLocation(Irp);
+    Status = IoCallDriver(Adapter->NdisMiniportBlock.NextDeviceObject, Irp);
+
+    IoDetachDevice(Adapter->NdisMiniportBlock.NextDeviceObject);
+        
+    RtlFreeUnicodeString(&Adapter->NdisMiniportBlock.MiniportName);
+
+    if (Adapter->NdisMiniportBlock.PnPDeviceState == NdisPnPDeviceStarted)
+    {
+        /* Remove adapter from adapter list for this miniport */
+        ExInterlockedRemoveEntryList(&Adapter->MiniportListEntry, &Adapter->NdisMiniportBlock.DriverHandle->Lock);
+
+        /* Remove adapter from global adapter list */
+        ExInterlockedRemoveEntryList(&Adapter->ListEntry, &AdapterListLock);
+
+        KeCancelTimer(&Adapter->NdisMiniportBlock.WakeUpDpcTimer.Timer);
+
+        Adapter->NdisMiniportBlock.DriverHandle->MiniportCharacteristics.HaltHandler(Adapter);
+    }
+
+    if (Adapter->NdisMiniportBlock.EthDB)
```

Comment:
Any way to factor things such that this function calls `NdisIPnPStopDevice`, since I would expect that to be a subset of the work that needs to be done for remove?